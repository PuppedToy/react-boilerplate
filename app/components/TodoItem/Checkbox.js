import styled from 'styled-components';
import PropTypes from 'prop-types';
import imgChecked from './checkbox-checked.png';
import imgUnchecked from './checkbox-unchecked.png';

const Checkbox = styled.div`
  width: 40px;
  height: 40px;
  overflow-y: auto;
  cursor: pointer;
  border-style: none;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  background-image: ${({ done }) => `url(${done ? imgChecked : imgUnchecked})`};

  &:first-child {
    border-top: none;
  }
`;

Checkbox.propTypes = {
  done: PropTypes.bool,
};

export default Checkbox;
