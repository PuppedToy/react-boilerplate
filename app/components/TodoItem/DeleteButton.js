import styled from 'styled-components';
import img from './delete.png';

const DeleteButton = styled.div`
  width: 35px;
  height: 35px;
  cursor: pointer;
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
  background-image: url(${img});
`;

export default DeleteButton;
