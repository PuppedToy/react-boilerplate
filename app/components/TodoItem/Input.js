import styled from 'styled-components';

const Input = styled.input`
  width: 100%;
  height: 40px;
  border: 0;
`;

export default Input;
