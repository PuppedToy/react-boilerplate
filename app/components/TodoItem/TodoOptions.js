import styled from 'styled-components';

const TodoOptions = styled.div`
  display: flex;
  align-items: center;
  margin-left: 20px;
`;

export default TodoOptions;
