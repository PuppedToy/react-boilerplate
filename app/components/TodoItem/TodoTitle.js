import styled from 'styled-components';

const TodoTitle = styled.div`
  overflow-x: auto;
  width: 100%;
`;

export default TodoTitle;
