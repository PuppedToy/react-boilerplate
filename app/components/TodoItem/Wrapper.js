import styled from 'styled-components';

const Wrapper = styled.div`
  width: 100%;
  height: 40px;
  white-space: nowrap;
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  border-bottom: 1px solid;
  margin-top: 10px;
  padding-left: 10px;
`;

export default Wrapper;
