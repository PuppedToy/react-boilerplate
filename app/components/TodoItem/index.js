/**
 *
 * TodoItem
 *
 */

import React from 'react';
import PropTypes from 'prop-types';

import Checkbox from './Checkbox';
import DeleteButton from './DeleteButton';
import Wrapper from './Wrapper';
import TodoOptions from './TodoOptions';
import TodoTitle from './TodoTitle';
import Input from './Input';

function TodoItem({
  id,
  title,
  done,
  focused,
  deleteTodo,
  focusTodo,
  editTodo,
  toggleTodo,
  onChangeTodoInput,
}) {
  const inputKeyDownHandler = ({ code, which }) => {
    if (editTodo && ((code && code === 13) || (which && which === 13)))
      editTodo();
  };
  const inputOnChangeHandler = ({ target }) => {
    if (
      onChangeTodoInput &&
      target &&
      Object.hasOwnProperty.call(target, 'value')
    )
      onChangeTodoInput(target.value);
  };
  const emptyHandler = () => {};

  return (
    <Wrapper id={`todo-${id}`}>
      {focused ? (
        <Input
          type="text"
          value={title}
          onChange={inputOnChangeHandler}
          onKeyDown={inputKeyDownHandler}
          data-testid={`editInput-${id}`}
        />
      ) : (
        <TodoTitle
          onClick={() => {
            if (focusTodo) focusTodo();
          }}
          onKeyDown={emptyHandler}
          data-testid={`todo-${id}`}
          type="button"
          role="button"
        >
          {title}
        </TodoTitle>
      )}
      <TodoOptions>
        <Checkbox
          data-testid={`toggle-${id}`}
          done={done}
          onClick={() => {
            if (toggleTodo) toggleTodo();
          }}
          onKeyDown={emptyHandler}
          type="button"
          role="button"
          alt={done ? 'uncheck' : 'check'}
        />
        <DeleteButton
          data-testid={`delete-${id}`}
          onClick={() => {
            if (deleteTodo) deleteTodo();
          }}
          onKeyDown={emptyHandler}
          type="button"
          role="button"
          alt="delete"
        />
      </TodoOptions>
    </Wrapper>
  );
}

TodoItem.propTypes = {
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  done: PropTypes.bool.isRequired,
  focused: PropTypes.bool.isRequired,
  deleteTodo: PropTypes.func,
  focusTodo: PropTypes.func,
  editTodo: PropTypes.func,
  toggleTodo: PropTypes.func,
  onChangeTodoInput: PropTypes.func,
};

export default TodoItem;
