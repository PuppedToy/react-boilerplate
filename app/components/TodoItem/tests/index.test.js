/**
 *
 * Tests for TodoItem
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import TodoItem from '../index';

describe('<TodoItem />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<TodoItem title="One item" id={1} done={false} focused={false} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should call delete hook when delete button is clicked', () => {
    const deleteTodoMock = jest.fn();
    const { getByTestId } = render(
      <TodoItem
        title="One item"
        id={1}
        done={false}
        focused={false}
        deleteTodo={deleteTodoMock}
      />,
    );

    fireEvent.click(getByTestId('delete'));
    expect(deleteTodoMock).toHaveBeenCalled();
  });

  it('Should call toggle hook when toggle button is clicked', () => {
    const toggleTodoMock = jest.fn();
    const { getByTestId } = render(
      <TodoItem
        title="One item"
        id={1}
        done={false}
        focused={false}
        toggleTodo={toggleTodoMock}
      />,
    );

    fireEvent.click(getByTestId('toggle'));
    expect(toggleTodoMock).toHaveBeenCalled();
  });

  it('Should call focus hook when clicked on the text', () => {
    const focusTodoMock = jest.fn();
    const title = 'One item';
    const { getByText } = render(
      <TodoItem
        title={title}
        id={1}
        done={false}
        focused={false}
        focusTodo={focusTodoMock}
      />,
    );

    fireEvent.click(getByText(title));
    expect(focusTodoMock).toHaveBeenCalled();
  });

  it('Should render an input when called with focused as true', async () => {
    const title = 'One item';
    const { getByTestId } = render(
      <TodoItem title={title} id={1} done={false} focused />,
    );

    expect(getByTestId('editInput')).toHaveValue(title);
  });

  it('Should call edit hook when pressed enter on edition mode', () => {
    const editTodoMock = jest.fn();
    const title = 'One item';
    const { getByTestId } = render(
      <TodoItem
        title={title}
        id={1}
        done={false}
        focused
        editTodo={editTodoMock}
      />,
    );

    fireEvent.keyDown(getByTestId('editInput'), {
      key: 'Enter',
      keyCode: 13,
    });

    expect(editTodoMock).toHaveBeenCalled();
  });

  it('Should call onChange hook when pressed any key edition mode', () => {
    const onChangeTodoInputMock = jest.fn();
    const title = 'One item';
    const { getByTestId } = render(
      <TodoItem
        title={title}
        id={1}
        done={false}
        focused
        onChangeTodoInput={onChangeTodoInputMock}
      />,
    );

    const input = getByTestId('editInput');
    const newValue = 'Changed value';

    fireEvent.change(input, { target: { value: newValue } });
    expect(onChangeTodoInputMock).toHaveBeenCalledWith(newValue);
  });

  it('Should do nothing when pressed enter and edition mode is disabled', () => {
    const editTodoMock = jest.fn();
    const title = 'One item';
    const {
      container: { firstChild },
    } = render(
      <TodoItem
        title={title}
        id={1}
        done={false}
        focused={false}
        editTodo={editTodoMock}
      />,
    );

    fireEvent.keyDown(firstChild, { key: 'Enter', keyCode: 13 });
    expect(editTodoMock).not.toHaveBeenCalled();
  });

  it('Should do nothing when pressed enter and input is not focused', () => {
    const editTodoMock = jest.fn();
    const title = 'One item';
    const {
      container: { firstChild },
    } = render(
      <TodoItem
        title={title}
        id={1}
        done={false}
        focused
        editTodo={editTodoMock}
      />,
    );

    fireEvent.keyDown(firstChild, { key: 'Enter', code: 13, which: 13 });
    expect(editTodoMock).not.toHaveBeenCalled();
  });

  // @TODO Further edition tests should be performed on TodoList, like checking if it actually selects
  // an item after triggering its focus todo or if it does unfocus an item after editing it

  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it.skip('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <TodoItem title="One item" id={1} done={false} focused={false} />,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
