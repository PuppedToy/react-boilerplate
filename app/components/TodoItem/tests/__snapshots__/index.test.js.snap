// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> Should render and match the snapshot 1`] = `
<div
  class="Wrapper-ziinx6-0 ecjWnr"
  id="todo-1"
>
  <div
    class="TodoTitle-xxpseh-0 cHcgZX"
    data-testid="todo-1"
    role="button"
    type="button"
  >
    One item
  </div>
  <div
    class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
  >
    <div
      alt="check"
      class="Checkbox-eu7a0t-0 hjQtxn"
      data-testid="toggle-1"
      role="button"
      type="button"
    />
    <div
      alt="delete"
      class="DeleteButton-sc-2zb8s8-0 ceLVod"
      data-testid="delete-1"
      role="button"
      type="button"
    />
  </div>
</div>
`;

exports[`<TodoItem /> Should render and match the snapshot when done is true 1`] = `
<div
  class="Wrapper-ziinx6-0 ecjWnr"
  id="todo-1"
>
  <input
    class="Input-tpeqbo-0 itMdwa"
    data-testid="editInput-1"
    type="text"
    value="One item"
  />
  <div
    class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
  >
    <div
      alt="uncheck"
      class="Checkbox-eu7a0t-0 hjQtxn"
      data-testid="toggle-1"
      role="button"
      type="button"
    />
    <div
      alt="delete"
      class="DeleteButton-sc-2zb8s8-0 ceLVod"
      data-testid="delete-1"
      role="button"
      type="button"
    />
  </div>
</div>
`;

exports[`<TodoItem /> Should render and match the snapshot when focused is true 1`] = `
<div
  class="Wrapper-ziinx6-0 ecjWnr"
  id="todo-1"
>
  <input
    class="Input-tpeqbo-0 itMdwa"
    data-testid="editInput-1"
    type="text"
    value="One item"
  />
  <div
    class="TodoOptions-sc-1s3b1hj-0 jWmTOe"
  >
    <div
      alt="check"
      class="Checkbox-eu7a0t-0 hjQtxn"
      data-testid="toggle-1"
      role="button"
      type="button"
    />
    <div
      alt="delete"
      class="DeleteButton-sc-2zb8s8-0 ceLVod"
      data-testid="delete-1"
      role="button"
      type="button"
    />
  </div>
</div>
`;
